<a href="https://colab.research.google.com/github/Clirmoo/Emerging-Technologies-1/blob/main/Copy_of_Activity_1_NumPy_and_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Technological Institute of the Philippines | Quezon City - Computer Engineering
--- | ---
Course Code: | CPE 018
Code Title: | Emerging Technologies in CpE 1 - Fundamentals of Computer Vision
1st Semester | AY 2025-2026
<hr> | <hr>
<u>**ACTIVITY NO. 1**</u> | <u>**NumPy and Pandas (Tutorial)**</u>
**Name** | Lim, Jusin Adrian D.
**Section** | CPE31S2
**Date Performed**: 07/29/2025 |
**Date Submitted**: 07/29/2025 |
**Instructor**: | Eng. Jimlord M. Quejado
<hr>

## 1. Objectives

This activity aims to familiarize students with popular data processing tools, such as NumPy and Pandas, and apply them on sample data.

## 2. Intended Learning Outcomes

After this activity, the student should be able to:

* Utilize pandas and numpy in fundamental data analysis.
* Demonstrate data analysis operations using numpy and pandas.

## 3. Procedures and Outputs

### 3.1 Introduction
In this activity, a look at two importal tools will be done before jumping into using opencv and more advanced applications. We are looking at NumPy and Pandas!

From Pandas' website:

> *Pandas is a very popular library for working with data (its goal is to be the most powerful and flexible open-source tool, and in our opinion, it has reached that goal). DataFrames are at the center of pandas. A DataFrame is structured like a table or spreadsheet. The rows and the columns both have indexes, and you can perform operations on rows or columns separately.*

> *A pandas DataFrame can be easily changed and manipulated. Pandas has helpful functions for handling missing data, performing operations on columns and rows, and transforming data. If that wasn’t enough, a lot of SQL functions have counterparts in pandas, such as join, merge, filter by, and group by. With all of these powerful tools, it should come as no surprise that pandas is very popular among data scientists.*

<hr/>

From NumPy's website:
> *NumPy is an open-source Python library that facilitates efficient numerical operations on large quantities of data. There are a few functions that exist in NumPy that we use on pandas DataFrames. For us, the most important part about NumPy is that pandas is built on top of it. So, NumPy is a dependency of Pandas.*

<hr/>

*Disclaimer: Parts of this activity are sourced from https://endaq.com/ and other open source tutorials for data processing tools.*

#### 3.1.1 Recap of Python Introduction

1. Python is popular for good reason
2. There are many ways to interact with Python
  * Here we are in Google Colab based on Jupyter Notebooks
3. There are many open source libraries to use
  * Today we are covering two of the most popular:
    * Numpy
    * Pandas

<img src="https://info.endaq.com/hubfs/20210928-python-ide-libraries-inforgraphic-FINAL-2.png" width="700">

#### 3.1.2 SO MANY Other Resources
Python is incredibly popular and so there are a lot of resources you can tap into online. You can either:

* Just start coding and google specific questions when you get stuck (my preference for learning)
* Follow a few online tutorials (like this first)
* Do both!

**Here are a few resources:**
* [Jake VanderPlas Python Data Science Handbook](https://colab.research.google.com/github/jakevdp/PythonDataScienceHandbook/blob/master/notebooks/Index.ipynb)
  * [Buy the book](https://www.amazon.com/Python-Data-Science-Handbook-Essential-ebook-dp-B01N2JT3ST/dp/B01N2JT3ST/ref=mt_other?_encoding=UTF8&me=&qid=) for $35
  * Go through a series of [well documented Colab Notebooks](https://colab.research.google.com/github/jakevdp/PythonDataScienceHandbook/blob/master/notebooks/Index.ipynb)
  * **Highly recommended resource!**
* [Code Academy: Analyze Data with Python](https://www.codecademy.com/learn/paths/analyze-data-with-python)
* [Udemy Academy: Data Analysis with Pandas & NumPy in Python for Beginner](https://www.udemy.com/course/python-data-analysis-2020/)
* [Datacamp: Introduction to Python](https://www.datacamp.com/courses/intro-to-python-for-data-science)


<img src="https://info.endaq.com/hubfs/datacamp-numpy-cheat-sheet.png" width="800">

<img src="https://info.endaq.com/hubfs/pandas-cheat-sheet1.png" width="800">

<img src="https://info.endaq.com/hubfs/pandas-cheat-sheet2.png" width="800">


### 3.2 NumPy

NumPy provides an in depth overview of [what exactly NumPy is](https://numpy.org/doc/stable/user/whatisnumpy.html). Quoting them:
>*At the core of the NumPy package, is the ndarray object. This encapsulates n-dimensional arrays of homogeneous data types, with many operations being performed in compiled code for performance.*

Let's get started by importing it, typically shortened to `np` because of how frequently it's called. This will come standard in most Python distributions such as Anaconda. If you need to install it simply:
~~~
!pip install numpy
~~~

In [ ]:
import numpy as np

#### 3.2.1 Creating Arrays

#### 3.2.2 Manually or from Lists
Manually create a list and demonstrate that operations on that list are difficult.

In [ ]:
lst = [0, 1, 2, 3]
lst * 2

[0, 1, 2, 3, 0, 1, 2, 3]

That isn't what we expected! For lists, we need to loop through all elements to apply a function to them which can be incredibly time consuming.

In [ ]:
[i * 2 for i in lst]

[0, 2, 4, 6]

Now lets make a numpy array and once in an array, let's show how intuitive operations are now that they are performed element by element.

In [ ]:
array = np.array(lst)
print(array)
print(array * 2)
print(array + 2)

[0 1 2 3]
[0 2 4 6]
[2 3 4 5]


Let's create a 2D matrix of 32 bit floats.

In [ ]:
np.array([[1, 0, 0],
          [0, 1, 0],
          [0, 0, 1]], dtype=np.float32)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]], dtype=float32)

#### 3.2.3 Using Functions

We'll go through a few here, but for more in depth examples and options see NumPy's [Array Creation Routines](https://numpy.org/doc/stable/reference/routines.array-creation.html). These first few examples are for very basic arrays/matrices.

In [ ]:
np.zeros(5)

array([0., 0., 0., 0., 0.])

In [ ]:
np.zeros([2, 3])

array([[0., 0., 0.],
       [0., 0., 0.]])

In [ ]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

Now let's start making sequences.

In [ ]:
np.arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
np.arange(0,  #start
          10) #stop (not included in array)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
np.arange(0,  #start
          10, #stop (not included in array)
          2)  #step size

array([0, 2, 4, 6, 8])

In [ ]:
np.linspace(0,  #start
            11, #stop (default to be included, can pass in endpoint=False)
            5)  #number of data points evenly spaced

array([ 0.  ,  2.75,  5.5 ,  8.25, 11.  ])

In [ ]:
np.logspace(0, #output starts being raised to this value
            4, #ending raised value
            4) #number of data points

array([1.00000000e+00, 2.15443469e+01, 4.64158883e+02, 1.00000000e+04])

Logspace is the equivalent of rasing a base by a linspaced array.

In [ ]:
10 ** np.linspace(0,3,4)

array([   1.,   10.,  100., 1000.])

In [ ]:
np.logspace(0, 4, 5, base=2)

array([ 1.,  2.,  4.,  8., 16.])

If you don't want to have to do the mental math to know what exponent to raise the values to, you can use geomspace but this only helps for base of 10.

In [ ]:
np.geomspace(1, 1000, 4)

array([   1.,   10.,  100., 1000.])

Random numbers!

In [ ]:
np.random.rand(5)

array([0.18962665, 0.23025895, 0.17167183, 0.69205744, 0.58853583])

In [ ]:
np.random.rand(2, 3)

array([[0.33514702, 0.48716713, 0.9618472 ],
       [0.96251775, 0.63552561, 0.01170064]])

#### 3.2.4 Indexing

Let's first create a simple array.

In [ ]:
array = np.arange(1, 11, 1)
array

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

Now index the first item.

In [ ]:
array[0]

np.int64(1)

Index the last item.

In [ ]:
array[-1]

np.int64(10)

Grab every 2nd item.

In [ ]:
array[::2]

array([1, 3, 5, 7, 9])

Grab every second item starting at index of 1 (the second value)

In [ ]:
array[1::2]

array([ 2,  4,  6,  8, 10])

Start from the second item, going to the 7th but skipping every other. (Our first time using the full `array[start (inclusive): stop (exclusive): step]` array 'slicing' notation)

In [ ]:
array[1:6:2]

array([2, 4, 6])

Reverse the order.

In [ ]:
array[::-1]

array([10,  9,  8,  7,  6,  5,  4,  3,  2,  1])

Boolean operations to index the array.

In [ ]:
array[array < 5]

array([1, 2, 3, 4])

In [ ]:
array[array * 2 < 5]

array([1, 2])

Integer list can also index arrays.

In [ ]:
array[[1,3,5]]

array([2, 4, 6])

Now let's create a slightly more complicated, 2 dimensional array.

In [ ]:
array_2d = np.arange(10).reshape((2, 5))
array_2d

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

Indexing the second dimension.

In [ ]:
array_2d[:, 1]

array([1, 6])

Any combination of these indexing methods works as well.

In [ ]:
array_2d[[True, False], 1::2]

array([[1, 3]])

In [ ]:
array_2d[1, [0,1,4]]

array([5, 6, 9])

#### 3.2.5 Operations

In [ ]:
array

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [ ]:
array + 100

array([101, 102, 103, 104, 105, 106, 107, 108, 109, 110])

In [ ]:
array * 2

array([ 2,  4,  6,  8, 10, 12, 14, 16, 18, 20])

To raise all the elements in an array to an exponent we have to use the notation `**` not `^`.

In [ ]:
array ** 2

array([  1,   4,   9,  16,  25,  36,  49,  64,  81, 100])

Use that shape to create a new array matching it to do operations with.

In [ ]:
array2 = np.arange(array.shape[0]) * 5
array2

array([ 0,  5, 10, 15, 20, 25, 30, 35, 40, 45])

In [ ]:
array2 + array

array([ 1,  7, 13, 19, 25, 31, 37, 43, 49, 55])

#### 3.2.6 Stats of an Array

In [ ]:
print(array)
print(array_2d)

[ 1  2  3  4  5  6  7  8  9 10]
[[0 1 2 3 4]
 [5 6 7 8 9]]


In [ ]:
print(array.shape)
print(array_2d.shape)

(10,)
(2, 5)


In [ ]:
print(len(array))
print(len(array_2d))

10
2


In [ ]:
print(array.max())
print(array_2d.max())

10
9


In [ ]:
print(array.min())
print(array_2d.min())

1
0


In [ ]:
array.std()

np.float64(2.8722813232690143)

In [ ]:
array.cumsum()

array([ 1,  3,  6, 10, 15, 21, 28, 36, 45, 55])

In [ ]:
array.cumprod()

array([      1,       2,       6,      24,     120,     720,    5040,
         40320,  362880, 3628800])

#### 3.2.7 Constants

In [ ]:
np.pi

3.141592653589793

In [ ]:
np.e

2.718281828459045

In [ ]:
np.inf

inf

#### 3.2.8 Functions

In [ ]:
np.sin(np.pi / 2)

np.float64(1.0)

In [ ]:
np.cos(np.pi)

np.float64(-1.0)

In [ ]:
np.log(np.e)

np.float64(1.0)

In [ ]:
np.log10(100)

np.float64(2.0)

In [ ]:
np.log2(64)

np.float64(6.0)

To demonstrate rounding, let's first make a new array with decimals.

In [ ]:
array = np.arange(4) / 3
print(array)
np.around(array, 2)

[0.         0.33333333 0.66666667 1.        ]


array([0.  , 0.33, 0.67, 1.  ])

#### 3.2.9 Looping vs Vectorization

As mentioned in the beginning, NumPy uses machine code with their ndarray objects which is what leads to the performance improvements. Let's demonstrate this by constructing a simple sine wave.

In [ ]:
fs = 500 #sampling rate in Hz
d_t = 1 / fs #time steps in seconds
n_step = 1000 #number of steps (there will be n_step+1 data points)

amp = 1 #amplitude of sine wave
f = 2 #frequency of sine wave

time = np.linspace(0,             #start time
                   n_step * d_t,    #end time
                   num=n_step + 1)  #number of data points, one more than the steps to include an endpoint
time.shape

(1001,)

First we make a function to loop through each element and calculate the amplitude.

In [ ]:
def sine_wave_with_loop(time, amp, f, phase=0):
  length = time.shape[0]
  wave = np.zeros(length)

  for i in range(length-1):
    wave[i] = np.sin(2 * np.pi * f * time[i] + phase * np.pi / 180)*amp
  return wave

Now let's time how quickly that executes for our time array of 1,001 data points.

In [ ]:
%timeit sine_wave_with_loop(time, amp, f)

1.64 ms ± 513 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


Now let's do the same using NumPy's sine function and vectorization.

In [ ]:
def sine_wave_with_numpy(time, amp, f, phase=0):
  """Takes in a time array and sine wave parameters, returns an array of the sine wave amplitude."""
  return np.sin(2 * np.pi * f * time + phase * np.pi / 180) * amp

Notice my docstrings!

In [ ]:
help(sine_wave_with_numpy)

Help on function sine_wave_with_numpy in module __main__:

sine_wave_with_numpy(time, amp, f, phase=0)
    Takes in a time array and sine wave parameters, returns an array of the sine wave amplitude.



In [ ]:
%timeit sine_wave_with_numpy(time, amp, f)

KeyboardInterrupt: 

Using vectorization is about **100x faster!** And this increases the longer the loops are.

#### 3.2.10 Why Vectorization Works So Much Faster
The above example highlights that NumPy is much faster, but why? Because it is using compiled machine code under the hood for it's operations.

Python has the [Numba](http://numba.pydata.org/) package which can be used to do this compilation which we will do to highlight just why NumPy is faster (and recommended!).

In [ ]:
from numba import njit

numba_sine_wave_with_loop = njit(sine_wave_with_loop)
numba_sine_wave_with_numpy = njit(sine_wave_with_numpy)

In [ ]:
%timeit numba_sine_wave_with_loop(time, amp, f)
%timeit numba_sine_wave_with_numpy(time, amp, f)

Let's combine this into a DataFrame we'll discuss next in more detail, but here's a preview.

In [ ]:
import pandas as pd

time_data = pd.DataFrame({'Time (us)':[2.77*100, 27.5, 23.1, 22.6],
                          'Method':['Loop','NumPy','Loop','NumPy'],
                          'Numba?':['w/o','w/o','w','w']}
)
time_data

Now let's plot it and preview Plotly!

In [ ]:
!pip install --upgrade -q plotly
import plotly.express as px

fig = px.bar(time_data,
             x="Method",
             y="Time (us)",
             color="Numba?",
             title="Compute Time of a Sine Wave for 1000 Elements",
             barmode='group')
fig.show()

### 3.3 Pandas
Pandas is built *on top of* NumPy meaning that the data is stored still as NumPy ndarray objects under the hood. But it exposes a much more intuitive labeling/indexing architecture and allows you to link arrays of different types (strings, floats, integers etc.) to one another.

To quote Jake VanderPlas:
>*At the very basic level, Pandas objects can be thought of as enhanced versions of NumPy structured arrays in which the rows and columns are identified with labels rather than simple integer indices.*

To start, import pandas as `pd`, again this will come standard in virtually all Python distributions such as Anaconda. But to install is simply:
~~~
!pip install pandas
~~~

In [ ]:
import pandas as pd

There are three types of Pandas objects, we'll only focus on the first two:
1. **Series** – 1D labeled homogeneous array, sizeimmutable
2. **Data Frames** – 2D labeled, size-mutable tabular structure with heterogenic columns
3. **Panel** – 3D labeled size mutable array.

#### 3.3.1 Creating a Series
First let's create a few numpy arrays.

In [ ]:
amplitude = sine_wave_with_numpy(time, amp, f, 180)
print(time)
print(amplitude)

Now let's see what a series looks like made from one of the arrays.

In [ ]:
pd.Series(amplitude)

This type of series has some value, but you really start to see it when you add in an index.

In [ ]:
series = pd.Series(data=amplitude,
                   index=time,
                   name='Amplitude')
series

Here's where Pandas shines - indexing is much more intuitive (and inclusive) to specify based on labels, not those confusing integer locations. We'll come back to this when we have the dataframe next too.

In [ ]:
series[0: 0.01]

Being able to plot quickly is also a plus!

In [ ]:
series.plot()

Remember we never left the NumPy array, it is still here and can be accessed with the following.

In [ ]:
series.values

In [ ]:
series.to_numpy()

#### 3.3.2 Creating a DataFrame

A DataFrame is basically a sequence of aligned series objects, and by aligned I mean they share a common index or label. This let's us mix and match types easily among other benefits.

First we'll start creating dataframes using what is called a "dictionary" with keys and values.

In [ ]:
df = pd.DataFrame({"Phase 0": sine_wave_with_numpy(time, amp, f, 00),
                   "Phase 90": sine_wave_with_numpy(time, amp, f, 90),
                   "Phase 180": sine_wave_with_numpy(time, amp, f, 180),
                   "Phase 270": sine_wave_with_numpy(time, amp, f, 270)},
                  index=time)
df

#### 3.3.3 Plotting (Preview)

Dataframes also wrap around Matplotlib to allow for plotting directly from the dataframe object itself. This can also be done from the Pandas Series object too like we showed earlier.

In [ ]:
df.plot()

In [ ]:
df['Max'] = df.max(axis=1)
df['Min'] = df.min(axis=1)
df

This will be the topic of the next webinar, plotting with Plotly!

Note that I need to install an upgraded version of Plotly in Colab because the default Plotly Express version doesn't work in Colab (but their more advanced graph objects does).

In [ ]:
!pip install --upgrade -q plotly
import plotly.express as px

In [ ]:
px.line(df).show()

#### 3.3.4 Load from CSV

This dataset was discussed in a blog on [vibration metrics and used bearing data as an example](https://blog.endaq.com/top-vibration-metrics-to-monitor-how-to-calculate-them).

Note you don't *have* to use a CSV. They have a lot of other file formats natively supported ([see full list](https://pandas.pydata.org/docs/user_guide/io.html)):
* hdf
* feather
* pickle

But I know everyone likes CSVs!

In [ ]:
df = pd.read_csv('https://info.endaq.com/hubfs/Plots/bearing_data.csv', index_col=0)
df

#### 3.3.5 Save CSV

Like reading data, there are a host of native formats we can save data from a dataframe. [See documentation](https://pandas.pydata.org/docs/reference/io.html).

In [ ]:
df.to_csv('bearing-data.csv')

#### 3.3.6 Simple Analysis

In [ ]:
df.describe()

In [ ]:
df.std()

In [ ]:
df.max()

Note that these built in Pandas functions are using NumPy to process and are the equivalent of doing the following.

In [ ]:
np.max(df)

In [ ]:
df.quantile(0.25)

In [ ]:
df['abs(max)'] = df.abs().max(axis=1)
df

#### 3.3.7 Indexing
Here is where indexing in Python gets a whole lot more intuitive! A dataframe with an index let's use index values (time in this case) to slice the dataframe, not rely on the nth element in the arrays.

In [ ]:
df[0: 0.05]

We can also use the same convention as before by adding in a step definition, in this case we'll grab every 100th point.

In [ ]:
df[0: 0.05: 100]

There are ways to use the integer based indexing if you so desire.

In [ ]:
df.iloc[0:10]

#### 3.3.8 Rolling
I love the [rolling method](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rolling.html) which allows for easy rolling window calculations, something you'll do frequently with time series data.

In [ ]:
df.rolling(n).max()[::n]

In [ ]:
px.line(df.rolling(n).max()[::n]).show()

#### 3.3.9 Datetime Data (Yay Finance!)

Let's use Yahoo Finance and stock data as a relatable example of data with datetimes.

In [ ]:
!pip install -q yfinance
import yfinance as yf

In [ ]:
df =  yf.download(["SPY", "AAPL", "MSFT", "AMZN", "GOOGL"],
                  start='2019-01-01',
                  end='2021-09-24')
df

Let's compare not the price, but the relative performance.

In [ ]:
df = df['Adj Close']
df = df / df.iloc[0]
df

In [ ]:
px.line(df).show()

The [rolling function](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rolling.html) will play very nicely with datetime data as shown here when I get the moving average over a 40 day period. And this can handle unevenly sampled date easily.

In [ ]:
px.line(df.rolling('40d').mean()).show()

Indexing with datetime data though will require a slightly extra step, but then it is easy.

In [ ]:
from datetime import date
start = date(2021, 4, 1)
end = date(2021, 4, 30)

In [ ]:
df[start:end]

In [ ]:
pd.date_range(start='1/1/2019', end='08/31/2021', freq='M')

In [ ]:
df.resample(rule='Q').max()

#### 3.3.10 Sorting & Filtering on Tabular Data
To highlight filtering in DataFrames, we'll use a dataset with a bunch of different columns/series of different types. This data was pulled directly from the enDAQ cloud API off some example recording files.

In [ ]:
df = pd.read_csv('https://info.endaq.com/hubfs/data/endaq-cloud-table.csv')
df

In [ ]:
df.columns

There's a lot of data here! So we'll focus on just a handful of columns and convert the time in seconds to a datetime object.

In [ ]:
df = df[['serial_number_id', 'file_name', 'file_size', 'recording_length', 'recording_ts',
         'accelerationPeakFull', 'psuedoVelocityPeakFull', 'accelerationRMSFull',
         'velocityRMSFull', 'displacementRMSFull', 'pressureMeanFull', 'temperatureMeanFull']].copy()

df['recording_ts'] = pd.to_datetime(df['recording_ts'], unit='s')
df = df.sort_values(by=['recording_ts'], ascending=False)
df

Filtering is made simple with boolean expressions that can be combined. There is also a method to sort_values by columns/series.

In [ ]:
mask = df.recording_ts > pd.to_datetime('2021-01-01')
df[mask].sort_values(by=['serial_number_id'], ascending=False)

In [ ]:
mask = (df.recording_ts > pd.to_datetime('2021-01-01')) & (df.accelerationPeakFull > 100)
df[mask].sort_values(by=['accelerationPeakFull'], ascending=False)

Another preview to plotly, but visualizing dataframes is made easy, even with mixed types.

In [ ]:
px.scatter(df,
           x="recording_ts",
           y="accelerationRMSFull",
           size="recording_length",
           color="serial_number_id",
           hover_name="file_name",
           log_y=True,
           size_max=60).show()

Plotly automatically made my colors a colorbar because I specified it based on a numeric value. If instead I change the type to string and replot, we'll see discrete series for each device.

In [ ]:
df['device'] = df["serial_number_id"].astype(str)

px.scatter(df,
           x="recording_ts",
           y="accelerationRMSFull",
           size="recording_length",
           color="device",
           hover_name="file_name",
           log_y=True,
           size_max=60).show()


## 4. Supplementary Activity

Download the dataset attached on the canvas course and answer the following:

1. Identify the column names
2. Identify the data types of the data
3. Display the total number of records
4. Display the first 20 records
5. Display the last 20 records
6. Change the Outcome column to Diagnosis
7. Create a new column Classification that display "Diabetes" if the value of outcome is 1 , otherwise "No Diabetes"
8. Create a new dataframe "withDiabetes" that gathers data with diabetes
9. Create a new dataframe "noDiabetes" thats gathers data with no diabetes
10. Create a new dataframe "Pedia" that gathers data with age 0 to 19
11. Create a new dataframe "Adult" that gathers data with age greater than 19
12. Use numpy to get the average age and glucose value.
13. Use numpy to get the median age and glucose value.
14. Use numpy to get the middle values of glucose and age.
15. Use numpy to get the standard deviation of the skinthickness.

In [ ]:
#1 Identify the column names
import pandas as pd
import numpy as np
df = pd.read_csv('diabetes.csv')
df.columns

#Observation: I already load the column names with the df.columns command.

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [ ]:
#2 Identify the data types of the data
df.dtypes

#Observation: with the df.types command I as able to see the data type of each lines.

Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object

In [ ]:
#3 Display the total number of records
len(df)

#Observation: with the len(df) function I was able to know the total length of the file.

768

In [ ]:
#4 Display the frist 20 records
df.head(20)

#Observation: With the df.head() function with the parameter of 20. I was able to display the first 20 records.

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0             6      148             72             35        0  33.6   
1             1       85             66             29        0  26.6   
2             8      183             64              0        0  23.3   
3             1       89             66             23       94  28.1   
4             0      137             40             35      168  43.1   
5             5      116             74              0        0  25.6   
6             3       78             50             32       88  31.0   
7            10      115              0              0        0  35.3   
8             2      197             70             45      543  30.5   
9             8      125             96              0        0   0.0   
10            4      110             92              0        0  37.6   
11           10      168             74              0        0  38.0   
12           10      139             80              0        0  27.1   
13            1      189             60             23      846  30.1   
14            5      166             72             19      175  25.8   
15            7      100              0              0        0  30.0   
16            0      118             84             47      230  45.8   
17            7      107             74              0        0  29.6   
18            1      103             30             38       83  43.3   
19            1      115             70             30       96  34.6   

    DiabetesPedigreeFunction  Age  Outcome  
0                      0.627   50        1  
1                      0.351   31        0  
2                      0.672   32        1  
3                      0.167   21        0  
4                      2.288   33        1  
5                      0.201   30        0  
6                      0.248   26        1  
7                      0.134   29        0  
8                      0.158   53        1  
9                      0.232   54        1  
10                     0.191   30        0  
11                     0.537   34        1  
12                     1.441   57        0  
13                     0.398   59        1  
14                     0.587   51        1  
15                     0.484   32        1  
16                     0.551   31        1  
17                     0.254   31        1  
18                     0.183   33        0  
19                     0.529   32        1

In [ ]:
#5 Display the last 20 records
df.tail(20)

#Observation: With the df.tail() function with the parameter of 20. I was able to display the last 20 records.

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
748            3      187             70             22      200  36.4   
749            6      162             62              0        0  24.3   
750            4      136             70              0        0  31.2   
751            1      121             78             39       74  39.0   
752            3      108             62             24        0  26.0   
753            0      181             88             44      510  43.3   
754            8      154             78             32        0  32.4   
755            1      128             88             39      110  36.5   
756            7      137             90             41        0  32.0   
757            0      123             72              0        0  36.3   
758            1      106             76              0        0  37.5   
759            6      190             92              0        0  35.5   
760            2       88             58             26       16  28.4   
761            9      170             74             31        0  44.0   
762            9       89             62              0        0  22.5   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
748                     0.408   36        1  
749                     0.178   50        1  
750                     1.182   22        1  
751                     0.261   28        0  
752                     0.223   25        0  
753                     0.222   26        1  
754                     0.443   45        1  
755                     1.057   37        1  
756                     0.391   39        0  
757                     0.258   52        1  
758                     0.197   26        0  
759                     0.278   66        1  
760                     0.766   22        0  
761                     0.403   43        1  
762                     0.142   33        0  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0

In [ ]:
#6 Change the outcome column to diagnosis
df.rename(columns={'Outcome':'Diagnosis'})

#Observation: This operation returns a new DataFrame with the renamed column, but it doesn't modify the original DataFrame df in place unless you add the inplace=True argument.

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Diagnosis  
0                       0.627   50          1  
1                       0.351   31          0  
2                       0.672   32          1  
3                       0.167   21          0  
4                       2.288   33          1  
..                        ...  ...        ...  
763                     0.171   63          0  
764                     0.340   27          0  
765                     0.245   30          0  
766                     0.349   47          1  
767                     0.315   23          0  

[768 rows x 9 columns]

In [ ]:
#7 Create a new column Classification that display "Diabetes" if the value of outcome is 1 , otherwise "No Diabetes"
df['Classification'] = np.where(df['Outcome'] == 1, 'Diabetes', 'No Diabetes')

#Observation: This code adds a new column called 'Classification' to the DataFrame. It then populates this column with the value 'Diabetes' if the 'Outcome' column is 1, and 'No Diabetes' otherwise.

In [ ]:
#8 Create a new dataframe "withDiabetes" that gathers data with diabetes
df_withDiabetes = df[df['Outcome'] == 1]

#Observation: This code creates a new DataFrame called df_withDiabetes. It does this by selecting only the rows from the original DataFrame df where the value in the 'Outcome' column is equal to 1, which represents patients with diabetes.

In [ ]:
#9 Create a new dataframe "noDiabetes" thats gathers data with no diabetes
df_noDiabetes = df[df['Outcome'] == 0]

#Observation: This code creates a new DataFrame called df_noDiabetes. It does this by selecting only the rows from the original DataFrame df where the value in the 'Outcome' column is equal to 0, which represents patients without diabetes.

In [ ]:
#10 Create a new dataframe "Pedia" that gathers data with age 0 to 19
df_Pedia = df[(df['Age'] >= 0) & (df['Age'] <= 19)]

#Observation: This code creates a new DataFrame called df_Pedia. It does this by selecting only the rows from the original DataFrame df where the value in the 'Age' column is between 0 and 19, inclusive.

In [ ]:
#11 Create a new dataframe "Adult" that gathers data with age greater than 19
df_Adult = df[df['Age'] > 19]

#Observation: This code creates a new DataFrame named df_Adult. It filters the original DataFrame df to include only the rows where the value in the 'Age' column is greater than 19, effectively creating a subset of the data for adults.

In [ ]:
#12 Use numpy to get the average age and glucose value.
average_age = np.mean(df['Age'])
average_glucose = np.mean(df['Glucose'])

print(f"Average Age: {average_age}")
print(f"Average Glucose: {average_glucose}")

#Observation: This code calculates the average values for the 'Age' and 'Glucose' columns in the DataFrame using NumPy's mean function. It then prints these calculated averages to the console.

Average Age: 33.240885416666664
Average Glucose: 120.89453125


In [ ]:
#13 Use numpy to get the median age and glucose value.
median_age = np.median(df['Age'])
median_glucose = np.median(df['Glucose'])

print(f"Median Age: {median_age}")
print(f"Median Glucose: {median_glucose}")

#Observation: This code calculates and prints the median values for the 'Age' and 'Glucose' columns from the DataFrame.

Median Age: 29.0
Median Glucose: 117.0


In [ ]:
#14 Use numpy to get the middle values of glucose and age.
median_glucose = np.median(df['Glucose'])
median_age = np.median(df['Age'])

print(f"Middle Glucose Value: {median_glucose}")
print(f"Middle Age Value: {median_age}")

#Observation: This code calculates and prints the median values for the 'Glucose' and 'Age' columns from the DataFrame.

Middle Glucose Value: 117.0
Middle Age Value: 29.0


In [ ]:
#15 Use numpy to get the standard deviation of the skinthickness.
std_skinthickness = np.std(df['SkinThickness'])
print(f"Standard Deviation of Skin Thickness: {std_skinthickness}")

#Observation: This code calculates and prints the standard deviation of the 'SkinThickness' column from the DataFrame.

Standard Deviation of Skin Thickness: 15.941828626496978


## 5. Summary, Conclusions and Lessons Learned

Based on the supplementary activity, here is a summary of the findings, conclusions, and lessons learned:

**Summary of Findings:**

*   The dataset contains information about 768 patients, with features like glucose levels, blood pressure, BMI, age, and whether they have diabetes.
*   The data types are appropriate for each column, with most being numerical (integers and floats).
*   The "Outcome" column was successfully renamed to "Diagnosis" for better clarity.
*   A new "Classification" column was created to provide a more descriptive label ("Diabetes" or "No Diabetes") based on the "Diagnosis" column.
*   The data was successfully filtered to create separate DataFrames for patients with and without diabetes, as well as for different age groups (Pedia and Adult).
*   Key statistics were calculated using NumPy:
    *   The average age of the patients is approximately 33.24 years.
    *   The average glucose level is approximately 120.89.
    *   The median age is 29 years, and the median glucose level is 117.
    *   The standard deviation of the skin thickness is approximately 15.94.

**Conclusions:**

The analysis reveals that the dataset is well-structured and suitable for basic data analysis and machine learning tasks. The average glucose level of the patients in this dataset is in the pre-diabetes range, which could indicate a population at risk. The median age being lower than the average age suggests that the age distribution is skewed towards younger individuals. The ability to filter and create subsets of the data allows for more targeted analysis of specific patient groups.

**Lessons Learned:**

This activity provided hands-on experience with fundamental data manipulation and analysis techniques using pandas and NumPy. We learned how to load, inspect, clean, and transform data, which are essential skills for any data-related task. The use of NumPy for statistical calculations demonstrated its efficiency and ease of use for numerical operations on DataFrames. The process of creating new columns and filtering data based on conditions highlighted the power of pandas for data wrangling. This exercise serves as a solid foundation for more advanced data analysis and machine learning applications.

<hr/>

***Proprietary Clause***

*Property of the Technological Institute of the Philippines (T.I.P.). No part of the materials made and uploaded in this learning management system by T.I.P. may be copied, photographed, printed, reproduced, shared, transmitted, translated, or reduced to any electronic medium or machine-readable form, in whole or in part, without the prior consent of T.I.P.*